### PR TITLE
Provide the user with more options for setting the to_bigquery config

### LINKED ISSUE #1654: Provide ability to set more options when outputting joined data to bq table.
**Is your feature request related to a problem? Please describe.**

We need to be able to set expiries for the tables, project.dataset.tableName, and also option for create vs replace.

**Describe the solution you'd like**

Ideally the user can optionally pass in a config object like QueryJobConfig. QueryJobConfig has many options such as `destination` which allows one to set the destination of table and name, but it doesn't seem like it can handle create/replace (it has create_disposition which offers some limited control) and expiry.

**Describe alternatives you've considered**

N/A. Willem and a few of us briefly discussed this and he played around with the idea of passing in and out sql queries. So one could modify the query and pass it back in.

### PR DESCRIPTION
**Pull request by codyjlin**
Signed-off-by: Cody Lin <codyl@twitter.com>

**What this PR does / why we need it**:
This PR gives the user more freedom to set configuration options for the `to_bigquery` call (on a BigQueryRetrievalJob following a `get_historical_features` call). For example, the user can specify a destination `project.dataset.tableName`, an expiration timestamp for the table, and also an option for replace/overwrite vs fail if already exists.

**Which issue(s) this PR fixes**:
Fixes #1654

**Does this PR introduce a user-facing change?**:

```release-note
* User should supply to_bigquery() with a bigquery.QueryJobConfig.
* User can also call to_sql() to get/modify the SQL directly (i.e. to set expiration_timestamp)
```

**References**
- [google.cloud.bigquery.job.QueryJobConfig](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.QueryJobConfig.html
)
- This PR builds on the design in https://github.com/feast-dev/feast/pull/1634 when to_bigquery() was first added

### CODE REVIEW

File: sdk/python/tests/test_historical_retrieval.py [hunk 1]
```diff
@@ -454,13 +459,26 @@ def test_historical_features_from_bigquery_sources(
         actual_bq_temp_table = bigquery.Client().get_table(bq_temp_table_path)
         assert actual_bq_temp_table.table_id == bq_temp_table_path.split(".")[-1]
 
+        # Config to overwrite an existing table should succeed
+        job_config = bigquery.QueryJobConfig(
+            destination=path, write_disposition="WRITE_TRUNCATE"
```
Comment (codyjlin) [inline sdk/python/tests/test_historical_retrieval.py:464]: Note: [write_disposition](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.QueryJobConfig.html#google.cloud.bigquery.job.QueryJobConfig.write_disposition) has the string options listed in the docs [here](https://cloud.google.com/bigquery/docs/reference/rest/v2/Job#JobConfigurationQuery.FIELDS.write_disposition).

File: sdk/python/feast/infra/offline_stores/bigquery.py [hunk 2]
```diff
@@ -224,19 +223,17 @@ def to_df(self):
         df = self.client.query(self.query).to_dataframe(create_bqstorage_client=True)
         return df
 
-    def to_bigquery(self, dry_run=False) -> Optional[str]:
+    def to_sql(self):
+        return self.query
```
Comment (codyjlin) [inline sdk/python/feast/infra/offline_stores/bigquery.py:227]: Note: there's no option for expiration_timestamp in the [QueryJobConfig](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.QueryJobConfig.html), so we should also give more granular control for the user to modify and run the SQL directly (like proposal 4 in https://github.com/feast-dev/feast/issues/1654#issuecomment-864218588). An example of this is in my fork [here](https://github.com/codyjlin/feast/blob/offline-to-bq/sdk/python/feast/infra/offline_stores/bigquery.py#L247-L256).

File: sdk/python/feast/infra/offline_stores/bigquery.py [hunk 3]
```diff
@@ -247,8 +244,8 @@ def _block_until_done():
         if bq_job.exception():
             raise bq_job.exception()
 
-        print(f"Done writing to '{path}'.")
-        return path
+        print(f"Done writing to '{job_config.destination}'.")
```
Comment (mavysavydav) [inline sdk/python/feast/infra/offline_stores/bigquery.py:247]: job_config might be None

File: sdk/python/tests/test_historical_retrieval.py [hunk 4]
```diff
@@ -439,12 +440,16 @@ def test_historical_features_from_bigquery_sources(
         )
 
         # Just a dry run, should not create table
-        bq_dry_run = job_from_sql.to_bigquery(dry_run=True)
+        job_config = bigquery.QueryJobConfig(dry_run=True)
```
Comment (mavysavydav) [inline sdk/python/tests/test_historical_retrieval.py:443]: let's not write tests for these for the time being. It won't be worth the work and protection when tests are about to undergo refactoring. Also this is within a test that'll run 6 times due to its parametrizations

File: sdk/python/tests/test_historical_retrieval.py (hunk 1 above)
Comment (mavysavydav) [reply sdk/python/tests/test_historical_retrieval.py:464]: Same as comment above, and we don't need to test for individual settings in QueryJobConfig. We can assume they work

File: sdk/python/tests/test_historical_retrieval.py (hunk 4 above)
Comment (mavysavydav) [reply sdk/python/tests/test_historical_retrieval.py:443]: I also think these kind of small changes don't need test cases generally

File: sdk/python/feast/infra/offline_stores/bigquery.py (hunk 2 above)
Comment (woop) [reply sdk/python/feast/infra/offline_stores/bigquery.py:228]: I think we are missing docstrings here (and for the other methods). Can you please add it to your method? Also, should there be a return type?

File: sdk/python/feast/infra/offline_stores/bigquery.py [hunk 5]
```diff
@@ -224,31 +224,38 @@ def to_df(self):
         df = self.client.query(self.query).to_dataframe(create_bqstorage_client=True)
         return df
 
-    def to_bigquery(self, dry_run=False) -> Optional[str]:
+    def to_sql(self):
+        return self.query
+
+    def to_bigquery(self, job_config=None) -> Optional[str]:
```
Comment (woop) [inline sdk/python/feast/infra/offline_stores/bigquery.py:230]: Can you please add a type here?

File: sdk/python/feast/infra/offline_stores/bigquery.py [hunk 6]
```diff
@@ -224,31 +224,51 @@ def to_df(self):
         df = self.client.query(self.query).to_dataframe(create_bqstorage_client=True)
         return df
 
-    def to_bigquery(self, dry_run=False) -> Optional[str]:
+    def to_sql(self) -> str:
+        """
+        Returns the SQL query that can be modified and run by the user to create the BQ table.
```
Comment (woop) [inline sdk/python/feast/infra/offline_stores/bigquery.py:229]:
```suggestion
        Returns the SQL query that will be executed in BigQuery to build the historical feature table.
```

File: sdk/python/feast/infra/offline_stores/bigquery.py [hunk 7]
```diff
@@ -224,31 +224,51 @@ def to_df(self):
...
+        """
+        return self.query
+
+    def to_bigquery(self, job_config: bigquery.QueryJobConfig = None) -> Optional[str]:
+        """
+        Uploads the results of the BigQueryRetrievalJob directly to BigQuery.
```
Comment (woop) [inline sdk/python/feast/infra/offline_stores/bigquery.py:235]:
```suggestion
        Triggers the execution of a historical feature retrieval query and exports the results to a BigQuery table.
```

File: sdk/python/feast/infra/offline_stores/bigquery.py [hunk 8]
```diff
@@ -224,31 +224,51 @@ def to_df(self):
...
+
+        Args:
+            job_config: A bigquery.QueryJobConfig to specify options like destination table, dry run, etc.
```
Comment (woop) [inline sdk/python/feast/infra/offline_stores/bigquery.py:238]:
```suggestion
            job_config: An optional bigquery.QueryJobConfig to specify options like destination table, dry run, etc.
```

File: sdk/python/feast/infra/offline_stores/bigquery.py [hunk 9]
```diff
@@ -224,31 +224,51 @@ def to_df(self):
...
+
+        Returns:
+            Returns either the destination table name, none if job_config.dry_run, or raises an exception if job fails
```
Comment (woop) [inline sdk/python/feast/infra/offline_stores/bigquery.py:241]:
```suggestion
            Returns the destination table name or returns None if job_config.dry_run is True.
```

**Review by woop — APPROVED** [review]

**Pull request merged**